File: ruby/sorting.rb
require 'benchmark'

# Quicksort algorithm implementation
def quicksort(arr)
  return arr if arr.length <= 1

  pivot = arr[arr.length / 2]
  left = arr.select { |x| x < pivot }
  middle = arr.select { |x| x == pivot }
  right = arr.select { |x| x > pivot }

  quicksort(left) + middle + quicksort(right)
end

# Verify if array is sorted
def sorted?(arr)
  (0...arr.length - 1).all? { |i| arr[i] <= arr[i + 1] }
end

# Run sorting benchmark with 100,000 random integers
def run_benchmark
  # Generate 100,000 random integers
  arr = Array.new(100_000) { rand(1_000_000) }

  # Measure sorting time
  start_time = Process.clock_gettime(Process::CLOCK_MONOTONIC)
  sorted_arr = quicksort(arr)
  end_time = Process.clock_gettime(Process::CLOCK_MONOTONIC)
  execution_time = ((end_time - start_time) * 1000).round(2)

  # Verify sorting correctness
  correct = sorted?(sorted_arr)

  puts "Test: Sorting (Quicksort)"
  puts "Array size: #{arr.length}"
  puts "Execution time: #{execution_time} ms"
  puts "Correctly sorted: #{correct}"
end

if __FILE__ == $0
  run_benchmark
end

File: rust/src/sorting.rs
use rand::Rng;
use std::time::Instant;

/// Quicksort algorithm implementation
fn quicksort(arr: Vec<i32>) -> Vec<i32> {
    if arr.len() <= 1 {
        return arr;
    }

    let pivot = arr[arr.len() / 2];
    let mut left = Vec::new();
    let mut middle = Vec::new();
    let mut right = Vec::new();

    for &x in &arr {
        if x < pivot {
            left.push(x);
        } else if x == pivot {
            middle.push(x);
        } else {
            right.push(x);
        }
    }

    let mut result = quicksort(left);
    result.extend(middle);
    result.extend(quicksort(right));
    result
}

/// Verify if array is sorted
fn is_sorted(arr: &[i32]) -> bool {
    arr.windows(2).all(|w| w[0] <= w[1])
}

/// Run sorting benchmark with 100,000 random integers
pub fn run_benchmark() {
    let mut rng = rand::thread_rng();
    
    // Generate 100,000 random integers
    let arr: Vec<i32> = (0..100000).map(|_| rng.gen_range(0..1000000)).collect();

    // Measure sorting time
    let start_time = Instant::now();
    let sorted_arr = quicksort(arr.clone());
    let execution_time = start_time.elapsed().as_millis();

    // Verify sorting correctness
    let correct = is_sorted(&sorted_arr);

    println!("Test: Sorting (Quicksort)");
    println!("Array size: {}", arr.len());
    println!("Execution time: {} ms", execution_time);
    println!("Correctly sorted: {}", correct);
}
